pkg/chunked/cache_linux.go

```diff
@@ -579,50 +724,89 @@ func byteSliceAsString(b []byte) string {
 }
 
 func findTag(digest string, cacheFile *cacheFile) (string, uint64, uint64) {
-	if len(digest) != cacheFile.digestLen {
+	binaryDigest, err := appendBinaryDigest([]byte{}, digest)
+	if err != nil {
+		return "", 0, 0
+	}
+	if len(binaryDigest) != cacheFile.digestLen {
 		return "", 0, 0
 	}
+	found, off, len := findBinaryTag(binaryDigest, cacheFile)
+	if found {
+		return digest, off, len
+	}
+	return "", 0, 0
+}
 
+func findBinaryTag(binaryDigest []byte, cacheFile *cacheFile) (bool, uint64, uint64) {
 	nElements := len(cacheFile.tags) / cacheFile.tagLen
 
 	i := sort.Search(nElements, func(i int) bool {
-		d := byteSliceAsString(cacheFile.tags[i*cacheFile.tagLen : i*cacheFile.tagLen+cacheFile.digestLen])
-		return strings.Compare(d, digest) >= 0
+		d := cacheFile.tags[i*cacheFile.tagLen : i*cacheFile.tagLen+cacheFile.digestLen]
+		return bytes.Compare(d, binaryDigest) >= 0
 	})
 	if i < nElements {
-		d := string(cacheFile.tags[i*cacheFile.tagLen : i*cacheFile.tagLen+len(digest)])
-		if digest == d {
+		d := cacheFile.tags[i*cacheFile.tagLen : i*cacheFile.tagLen+cacheFile.digestLen]
+		if bytes.Equal(binaryDigest, d) {
 			startOff := i*cacheFile.tagLen + cacheFile.digestLen
-			parts := strings.Split(string(cacheFile.tags[startOff:(i+1)*cacheFile.tagLen]), "@")
 
-			off, _ := strconv.ParseInt(parts[0], 10, 64)
+			// check for corrupted data, there must be 2 u64 (off and len) after the digest.
+			if cacheFile.tagLen < cacheFile.digestLen+16 {
+				return false, 0, 0
+			}
+
+			offsetAndLen := cacheFile.tags[startOff : (i+1)*cacheFile.tagLen]
 
-			len, _ := strconv.ParseInt(parts[1], 10, 64)
-			return digest, uint64(off), uint64(len)
+			off := binary.LittleEndian.Uint64(offsetAndLen[:8])
+			len := binary.LittleEndian.Uint64(offsetAndLen[8:16])
+
+			return true, off, len
 		}
 	}
-	return "", 0, 0
+	return false, 0, 0
 }
 
 func (c *layersCache) findDigestInternal(digest string) (string, string, int64, error) {
 	if digest == "" {
 		return "", "", -1, nil
 	}
 
+	binaryDigest, err := appendBinaryDigest([]byte{}, digest)
+	if err != nil {
+		return "", "", 0, err
+	}
+
 	c.mutex.RLock()
 	defer c.mutex.RUnlock()
 
 	for _, layer := range c.layers {
-		digest, off, tagLen := findTag(digest, layer.cacheFile)
-		if digest != "" {
-			position := string(layer.cacheFile.vdata[off : off+tagLen])
-			parts := strings.SplitN(position, ":", 3)
-			if len(parts) != 3 {
-				continue
+		if !layer.cacheFile.bloomFilter.maybeContains(binaryDigest) {
+			continue
+		}
+		found, off, tagLen := findBinaryTag(binaryDigest, layer.cacheFile)
+		if found {
+			if uint64(len(layer.cacheFile.vdata)) <= off+tagLen {
+				return "", "", 0, err
```

Comment:
In all the added bound checks: This returns `"", "", 0, nil`, AFAICS that's considered a “success, found”. It should either contain an error value, or a -1 offset to indicate “not found”.

(And if the intent is to return `nil`, I’d find an explicit `nil` more readable.)